docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1438 @@
+trainer container.
+
+```yaml
+apiVersion: kubeflow.org/v2alpha1
+kind: TrainJob
+metadata:
+  name: pytorch-distributed
+  namespace: tenant-alpha
+spec:
+  trainingRuntimeRef:
+    name: pytorch-distributed-gpu
+  trainer:
+    image: docker.io/custom-training
+  podSpecOverrides:
+    - targetReplicatedJobs:
+        - initializer
+          node
+      containers:
+        - name: user-identity
+          value: 123
+        - name: trainer
+          volumeMounts:
+            - name: user-123-volume
+              mountPath: /workspace
+      volumes:
+        - name: user-123-volume
+          persistentVolumeClaim:
+            claimName: user-123-volume
+```
+
+## The Training Runtime API
+
+The `TrainingRuntime` is the pre-created configurations of model training on the cluster,
+representing as blueprints. For example, Elastic PyTorch training, MPI DeepSpeed configuration,
+BERT LLM Fine-Tuning.
+
+These blueprints can be deployed within the Training Operator control plane and stored in a Kubeflow
+public repository that users can apply to their clusters.
+
+Platform or ML engineers can tweak existing blueprints, based on their requirements. For example,
+using custom configurations.
+
+The Kubeflow Training Operator can maintain more Training Runtimes when the community is ready to
+support them. For example, runtimes for [Jax](https://jax.readthedocs.io/en/latest/index.html) or
+[MLX](https://ml-explore.github.io/mlx/build/html/index.html). We will support PyTorch and MPI runtimes.
+After initial implementation, we will support TensorFlow, XGboost, and PaddlePaddle runtimes, but
+it is out of scope for this KEP.
+
+The `TrainingRuntime` is immutable, and so to make a change, a new version of the `TrainingRuntime`
+must be created and then the user must change the `TrainJob` to point to the new version.
+This provides control as to how changes to runtimes propagate to existing training jobs.
+For example, when training is running for a long time (e.g. 1-2 months).
+
+In the future implementation, we will introduce a revision control mechanism similar to
+[Kubernetes Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#updating-a-deployment)
+to control versions of `TrainingRuntime` and enable rolling updates.
+
+We are going to create two CRDs: `TrainingRuntime` and `ClusterTrainingRuntime`. These runtimes have
+exactly the same APIs, but the first one is the namespace-scoped, the second is the cluster-scoped.
+If `trainingRuntimeRef` in `TrainJob` has the namespace, controller will use the `TrainingRuntime`,
+otherwise it will use the `ClusterTrainingRuntime`.
+
+```golang
+type TrainingRuntime struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Framework specific parameters.
+    MLSpec *MLSpec `json:"mlSpec,omitempty"`
+
+    // Number of nodes to execute training.
+    NumNodes int `json:"numNodes,omitempty"`
```

Comment:
```suggestion
    NumNodes *int32 `json:"numNodes,omitempty"`
```

This is optional field, right?

Reply:
Actually, we should set it to 1 by default, isn't ?

Reply:
> we should set it to 1 by default, isn't ?

That's right. If users do not set this field (numNodes == nil), we can default to 1.